### PR TITLE
feat(format):parsing and formatting markdown table 

### PR DESCRIPTION
**Pull request by Monchi**
## Summary

This PR implements parsing and formatting markdown table (part of #985). The AST structure is inspired by mdast.

### Side changes

- Rename `Strong` and `Emphasis` from "inline" to "textwrapping". "inline" should include others such as reference syntax. I'm not native english speaker and "textwrapping" maybe not a good choice. Please tell me if you have any alternative!

- Add `forceFormat` option to ignore file handler's capability on test.

## Test Plan

Input:
```markdown
|head 1|head 2|
| with space |without space|
|bottom 1|bottom 2|
```

Output:
```markdown
|   head 1   |   head 2    |
| with space |without space|
|  bottom 1  |  bottom 2   |
```

### CODE REVIEW

**Comment by ematipico** [comment]
Hi @Monchi ,

Thank you very much for this fantastic contribution!

About the naming convention, we tries to follow the GitHub spec and here they call it "inline" 

https://github.github.com/gfm/#inlines

**Comment by Monchi** [comment]
@ematipico 

Thank you very much for developing Rome continuously!

I totally agree on following the GitHub spec.

Currently `parseInline` parses only strong and emphasis. But according to the GitHub spec, "inline" also includes link reference and more.

So I renamed current `parseInline` to `parseTextWrapping` and implemented new `parseInline` to parse true "inline".

File: internal/markdown-parser/test-fixtures/tables/input.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+|head 1|head 2|
+| with space |ddadada|
+|**bold**|italic|
```
Comment (ematipico) [inline internal/markdown-parser/test-fixtures/tables/input.md:3]: The actual italic is missing :)
```suggestion
|**bold**|_italic_|
```

File: internal/markdown-parser/parser/paragraph.ts [hunk 2]
```diff
@@ -18,64 +17,23 @@ export function parseParagraph(
...
-						if (unknownToken.type === "OpenSquareBracket") {
-							return parseReference(parser);
-						}
-						return parseText(parser);
-					},
-				);
-				if (nodes) {
-					children.push(nodes);
-				}
+		if (isTable && token.type === "TablePipe") {
```
Comment (ematipico) [inline internal/markdown-parser/parser/paragraph.ts:20]: In theory we should not encounter `TablePipe` inside a paragraph. We should encounter just simple text (the pipe inside a paragraph is treated as text)

File: internal/markdown-parser/parser/table.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
+import {parseInline} from "@internal/markdown-parser/parser/inline";
+import {
+	AnyMarkdownInlineNode,
+	MarkdownTableCell,
+	MarkdownTableRow,
+} from "@internal/ast";
+
+import {MarkdownTable} from "./../../ast/markdown/tables/MarkdownTable";
```
Comment (ematipico) [inline internal/markdown-parser/parser/table.ts:8]: Wrong import, it should be imported from `@internal/ast`

File: internal/formatter/test-fixtures/markdown/tables/input.test.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+|head 1|head 2|
+| with space |without space|
+|bottom 1|bottom 2|
+
+```
+
+### `Output`
+
+```md
+|   head 1   |   head 2    |
```
Comment (ematipico) [inline internal/formatter/test-fixtures/markdown/tables/input.test.md:26]: I like the formatting but I have mixed feelings about the justifying the text. I am totally in favour of filling the table with spaces, so it looks neat. What do you think?

File: internal/formatter/builders/markdown/tables/MarkdownTableRow.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import {MarkdownTableRow} from "@internal/ast";
+import {Builder, Token} from "@internal/formatter";
+
+export default function MarkdownTableRow(
+	builder: Builder,
+	node: MarkdownTableRow,
+): Token {
+	throw Error("Must be rendered through MarkdownTable.");
```
Comment (ematipico) [inline internal/formatter/builders/markdown/tables/MarkdownTableRow.ts:8]: Is there a reason why you decided to use `MarkdownTable` to format this node?

**Review by ematipico — CHANGES_REQUESTED** [review]
Thank you for your contribution, lot of nice work done here! I left some feedback and question.

There's a couple of things I don't see:
- table headings https://github.github.com/gfm/#example-198
- handling of the text alignment https://github.github.com/gfm/#example-199

Do you plan to add them later with another PR?

Also, I think we should have more cases inside our tests. The only case I can see here is a "happy" case.

What about cases like this?

```
| Something
```

```
| Something | Else |
| Wow |
| Wow 
```

Also, the requires an header (`|----|`). Without the header, the table is rendered as text. I think we should consider this case.

**Comment by Monchi** [comment]
@ematipico Thank you for reviewing!
I had some misunderstandings about GFM spec. I'll add some "unhappy" test cases and update parser.

BTW, I think we need too long time to stabilize our full-scratch parser.
How do you think about including [micromark](https://github.com/micromark/micromark/) as a internal? We can generate [mdast](https://github.com/syntax-tree/mdast) with micromark, then convert it to rome's AST.

**Comment by ematipico** [comment]
I wasn't aware of `mdast` and we can definitely take inspiration from their categorisation and AST. Without realizing it, I think we already have some of their AST inside ours. 

About `micromark`, with Rome we want to try to do things ourselves without relying on third party packages. We can surely copy and credit third party code (we've done this already). About our markdown parser, we already implemented our logic, I'm not sure how it would play inside Rome code base. I didn't do any research. What do you think?

**Comment by Monchi** [comment]
@ematipico
I think it's good to copy `micromark` **for now**.
The important point is that we won't use `mdast` directly but use Rome AST that converted from `mdast`. So we can rewrite our parser later without affecting linter, formatter, etc. We already do this in JavaScript parser with babel-parser.

IMO, we should focus on making Rome ready for use to gather more contributors.

**Comment by ematipico** [comment]
Makes sense to me! Feel free to look into it if you have the time

**Comment by sebmck** [comment]
Just wanted to leave a quick note around copying from other projects. We want to be very careful with licensing, and API copyright is a real area of ambiguity. Preferably we should take inspiration but never straight up copy. There's almost always room for improvement.

**Comment by jamiebuilds** [comment]
Hi! Thank you so much for your contribution!

We recently made the decision to [migrate the Rome codebase to Rust](https://rome.tools/blog/2021/09/21/rome-will-be-rewritten-in-rust) and making some significant changes to its architecture.

We're going through and closing pull requests and issues that were specific to the TypeScript codebase, including this one. But while we're closing this, we'd love to have your involvement in the future.

Thanks again!

**Pull request closed without merging**